src/engine/Account-Imap.cpp

```diff
@@ -0,0 +1,942 @@
+
+    auto acc_val = new KvpValue(const_cast<GncGUID*>(xaccAccountGetGUID(acc)));
+
+    xaccAccountBeginEdit (m_account);
+
+    if (!category)
+        acc_slots->set_path({IMAP_FRAME, key}, acc_val);
+    else
+        acc_slots->set_path({IMAP_FRAME, category, key}, acc_val);
+
+    qof_instance_set_dirty (QOF_INSTANCE (m_account));
+    xaccAccountCommitEdit (m_account);
+
+    LEAVE(" ");
+}
+
+void
+GncImportMatchMap::delete_account (const char *category, const char *key)
+{
+    std::string kvp_path;
+    std::string delim = "/";
+
+    if (!key) return;
+
+    ENTER(" ");
+
+    std::string str_key (key);
+
+    if (!category)
+        kvp_path = IMAP_FRAME + delim + str_key;
+    else
+    {
+        std::string str_category (category);
+        kvp_path = IMAP_FRAME + delim + str_category + delim + str_key;
+    }
+    xaccAccountBeginEdit (m_account);
+
+    if (qof_instance_has_slot (QOF_INSTANCE (m_account), kvp_path.c_str()))
+    {
+        qof_instance_slot_delete (QOF_INSTANCE (m_account), kvp_path.c_str());
+
+        if (category)
+        {
+            std::string str_category (category);
+            kvp_path = IMAP_FRAME + delim + str_category;
+
+            qof_instance_slot_delete_if_empty (QOF_INSTANCE (m_account), kvp_path.c_str());
+        }
+        qof_instance_slot_delete_if_empty (QOF_INSTANCE (m_account), IMAP_FRAME);
+    }
+    qof_instance_set_dirty (QOF_INSTANCE (m_account));
+    xaccAccountCommitEdit (m_account);
+
+    LEAVE(" ");
+}
+
+
+struct account_token_count
+{
+    std::string account_guid; // returned account guid
+    int64_t token_count;      // occurances of a given token for this account_guid
+};
+
+/** total_count and the token_count for a given account let us calculate the
+ * probability of a given account with any single token
+ */
+struct token_accounts_info
+{
+    GList *accounts;           // array of struct account_token_count
+    int64_t total_token_count; // the sum of token_count for a given token
+};
+
+/** gpointer is a pointer to a struct token_accounts_info
+ * \note Can always assume that keys are unique, reduces code in this function
+ */
+static void
+buildTokenInfo(const char *key, const GValue *value, gpointer data)
+{
+    struct token_accounts_info *tokenInfo = (struct token_accounts_info*)data;
+    struct account_token_count* this_account;
+
+    //  PINFO("buildTokenInfo: account '%s', token_count: '%" G_GINT64_FORMAT "'", (char*)key,
+    //                  g_value_get_int64(value));
+
+    /* add the count to the total_token_count */
+    tokenInfo->total_token_count += g_value_get_int64(value);
+
+    /* allocate a new structure for this account and it's token count */
+    this_account = (struct account_token_count*)
+                   new (struct account_token_count);
+
+    /* fill in the account guid and number of tokens found for this account */
+    this_account->account_guid = ((char*)key);
+    this_account->token_count = g_value_get_int64(value);
```

Comment:
You shouldn't need this cast. If you do it's to cast away constness and you should make ```account_token_count::account_guid``` ```const std::string``` so that you're not. It should probably be ```const``` anyway, it's invariant.